Update sample_distorted_bounding_box to use Tensor Input for min_object_covered

yongtang opened the pull request:
This fix tries to address the issue raised in #10715 where it was not possible
to use dynamic values for min_object_covered in sample_distorted_bounding_box.

This fix created an Op of SampleDistortedBoundingBoxV2 so that min_object_covered
takes input. SampleDistortedBoundingBox remains the same (min_object_covered as attr).

Additional test cases have been added.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

yifeif commented:
@vrv could you take a look at this change? Thanks!

vrv commented on tensorflow/core/kernels/sample_distorted_bounding_box_op.cc:
```diff
@@ -275,6 +277,22 @@ class SampleDistortedBoundingBoxOp : public OpKernel {
+          errors::InvalidArgument("min_object_covered must be 0-D, got shape ",
+                                  min_object_covered.shape().DebugString()));
+
+      min_object_covered_ = min_object_covered.scalar<float>()();
```
Compute() can be run by multiple threads, so this is going to result in a potential memory race.  I would suggest doing something like:

float min_object_recovered_val = 0.0;
if (context->num_inputs() == 3) {
  ...
  min_object_recovered_val = min_object_recovered.scalar<float>()();
} else {
  min_object_recovered_val = min_object_recovered_;
}

and then use that float below.

vrv reviewed (changes requested):
Looks generally good!  I suspect in the future we can make the default python version of sample_distorted_bounding_box to use the v2 version, but we have to wait a few weeks to do that (for forward compatibility).

yongtang commented:
@vrv Thanks for the review. The PR has been updated. Please take a look and let me know if there are any issues.

vrv commented:
@tensorflow-jenkins test this please!

drpngx commented:
I think this adds an op and will need API review. Could you run:

```
   $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```

and push again? It'll be clear what the API change is.

yongtang commented:
Thanks @drpngx. The PR has been updated to address the api compatibility test failure. Please take a look and let me know if there are any issues.

drpngx commented:
Ugh, somehow all the commits ended up on the CL. Could you rebase and push again?

drpngx commented:
Adding @martinwicke for API review.

yongtang committed:
Update sample_distorted_bounding_box to use Tensor Input for min_object_covered

This fix tries to address the issue raised in #10715 where it was not possible
to use dynamic values for min_object_covered in sample_distorted_bounding_box.

This fix created an Op of SampleDistortedBoundingBoxV2 so that min_object_covered
takes input. SampleDistortedBoundingBox remains the same (min_object_covered as attr).

Additional test cases have been added.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update API golden to pass api_compatibility_test

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @drpngx . The PR has been rebased and pushed.

drpngx commented:
Jenkins, test this please.

martinwicke commented on tensorflow/tools/api/golden/tensorflow.image.pbtxt:
```diff
@@ -173,6 +173,10 @@ tf_module {
   member_method {
+    name: "sample_distorted_bounding_box_v2"
+    argspec: "args=[\'image_size\', \'bounding_boxes\', \'min_object_covered\', \'seed\', \'seed2\', \'aspect_ratio_range\', \'area_range\', \'max_attempts\', \'use_image_if_no_bounding_boxes\', \'name\'], varargs=None, keywords=None, defaults=[\'None\', \'None\', \'None\', \'None\', \'None\', \'None\', \'None\'], "
+  }
```
Please hide this op, as well as the original, and write a python wrapper which calls only the V2 version.

vrv replied:
Forwards compatibility requirements suggests we need to wait several weeks before changing python to use V2.

drpngx replied:
Aww, right. We could still hide the op, then switch to V2 in a couple of weeks. @yongtang if you're up for it, you'd need to change https://www.github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/hidden_ops.txt#L159 and write a wrapper like this: http://google3/third_party/tensorflow/python/ops/image_ops_impl.py?l=925

yongtang commented:
@martinwicke @vrv @drpngx The PR has been updated as suggested. Please take a look.

martinwicke commented on tensorflow/python/ops/image_ops.py:
```diff
@@ -62,6 +62,7 @@
 @@non_max_suppression
 @@non_max_suppression_v2
 @@sample_distorted_bounding_box
+@@sample_distorted_bounding_box_v2
```
Please don't expose the _v2 op directly. We need to register both kernels, but then only expose a single python op, which only generates the _v2 op. To do that, you currently have to mark both the v1 and v2 ops "hidden" in the BUILD file, and then write a python wrapper which calls the generated_image_ops._..._v2() directly (ignoring the linter that will complain about a protected symbol access).

We're working on improving this workflow but that's what it is for now.

martinwicke replied:
In fact, it's ever worse: Because we're guaranteeing forward compatibility, this change has to be split in two: One that introduces the new _v2 kernel (hidden), and no less than three weeks later, one that switches the behavior of the python op from v1 to _v2 by hiding the v1 op and adding the python wrapper.

drpngx replied:
remove

yongtang replied:
Thanks. Done.

yongtang commented:
@martinwicke Thanks for the review. There was some confusion from my side before about the workflow. Now the PR has been updated with the following changes:
This PR:
1. New kernel (`SampleDistortedBoundingBoxV2Op`) has been registered.
2. `SampleDistortedBoundingBoxV2` is hidden in python (tensorflow/python/ops/hidden_ops.txt)

Next PR (I will submit it after 3 weeks):
1. `SampleDistortedBoundingBox` will be hidden in python (tensorflow/python/ops/hidden_ops.txt)
2. An wrapper `sample_distorted_bounding_box` will be added in `image_ops_impl.py` and calls `sample_distorted_bounding_box_v2`.

Please take a look at the update PR and let me know if the above changes matches the API workflow.

yongtang commented:
@martinwicke @drpngx Thanks. The PR has been updated with both `SampleDistortedBoundingBox` and `SampleDistortedBoundingBoxV2` hidden. A wrapper has been added in image_ops_impl.py.

I will create another PR after 3 weeks. 

Please take a look.

drpngx commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -1472,3 +1473,104 @@ def total_variation(images, name=None):
+  """
+
+  with ops.name_scope(name, 'sample_distorted_bounding_box'):
+    return gen_image_ops._sample_distorted_bounding_box_v2(image_size,
```
You need the v1 version, then we'll switch this one to V2 in three weeks.

yongtang replied:
Done.

drpngx commented on tensorflow/core/kernels/sample_distorted_bounding_box_op.cc:
```diff
@@ -399,11 +420,15 @@ class SampleDistortedBoundingBoxOp : public OpKernel {
+  REGISTER_KERNEL_BUILDER(Name("SampleDistortedBoundingBox")    \
+                              .Device(DEVICE_CPU)               \
+                              .TypeConstraint<type>("T"),       \
+                          SampleDistortedBoundingBoxV2Op<type>) \
```
I would favor keeping the old code around (the whole class).

vrv replied:
As long as the v1 API contract is held, I feel like it doesn't matter how the OpKernel is written, and less code is better.  But my risk tolerance might be higher than everyone's :)

drpngx commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -1030,16 +1120,16 @@ method: A string specifying the interpolation method. Only 'bilinear' is
-  .Output("selected_indices: int32")
-  .Attr("iou_threshold: float = 0.5")
-  .SetShapeFn([](InferenceContext* c) {
+    .Input("boxes: float")
```
keep old 2-spaces identation, as per our style guide.

yongtang replied:
Done.

drpngx reviewed (changes requested):
@martinwicke or @vrv can chime in about keeping the old code around.

drpngx commented:
Jenkins, test this please.
